<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Стиль-заголовков" data-toc-modified-id="Стиль-заголовков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Стиль заголовков</a></span></li><li><span><a href="#Пропуски-значений" data-toc-modified-id="Пропуски-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Пропуски значений</a></span><ul class="toc-item"><li><span><a href="#Изучение-столбцов-с-пропусками" data-toc-modified-id="Изучение-столбцов-с-пропусками-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Изучение столбцов с пропусками</a></span></li><li><span><a href="#Обработка-столбца-'days_employed'" data-toc-modified-id="Обработка-столбца-'days_employed'-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Обработка столбца 'days_employed'</a></span></li><li><span><a href="#Обработка-столбца-'total_income'" data-toc-modified-id="Обработка-столбца-'total_income'-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Обработка столбца 'total_income'</a></span></li></ul></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span><ul class="toc-item"><li><span><a href="#Неявные-дубликаты" data-toc-modified-id="Неявные-дубликаты-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Неявные дубликаты</a></span></li><li><span><a href="#Явные-дубликаты" data-toc-modified-id="Явные-дубликаты-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Явные дубликаты</a></span><ul class="toc-item"><li><span><a href="#Обработка-столбцов-'family_status'-и-'family_status_id'" data-toc-modified-id="Обработка-столбцов-'family_status'-и-'family_status_id'-2.3.2.1"><span class="toc-item-num">2.3.2.1&nbsp;&nbsp;</span>Обработка столбцов 'family_status' и 'family_status_id'</a></span></li><li><span><a href="#Обработка-столбца-'purpose'" data-toc-modified-id="Обработка-столбца-'purpose'-2.3.2.2"><span class="toc-item-num">2.3.2.2&nbsp;&nbsp;</span>Обработка столбца 'purpose'</a></span></li><li><span><a href="#Удаление-лишних-данных" data-toc-modified-id="Удаление-лишних-данных-2.3.2.3"><span class="toc-item-num">2.3.2.3&nbsp;&nbsp;</span>Удаление лишних данных</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Проведение-исследования" data-toc-modified-id="Проведение-исследования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проведение исследования</a></span><ul class="toc-item"><li><span><a href="#Изучение-статистики" data-toc-modified-id="Изучение-статистики-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Изучение статистики</a></span></li><li><span><a href="#Составление-сводных-таблиц" data-toc-modified-id="Составление-сводных-таблиц-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Составление сводных таблиц</a></span><ul class="toc-item"><li><span><a href="#Сводная-таблица-по-наличию-детей" data-toc-modified-id="Сводная-таблица-по-наличию-детей-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Сводная таблица по наличию детей</a></span></li><li><span><a href="#Сводная-таблица-по-семейному-положению" data-toc-modified-id="Сводная-таблица-по-семейному-положению-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Сводная таблица по семейному положению</a></span></li><li><span><a href="#Сводная-таблица-по-уровню-дохода" data-toc-modified-id="Сводная-таблица-по-уровню-дохода-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Сводная таблица по уровню дохода</a></span></li><li><span><a href="#Сводная-таблица-по-целям-кредитов" data-toc-modified-id="Сводная-таблица-по-целям-кредитов-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Сводная таблица по целям кредитов</a></span></li></ul></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Обнаруженные-проблемы" data-toc-modified-id="Обнаруженные-проблемы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Обнаруженные проблемы</a></span></li></ul></div>

# Исследование надежности заёмщика

Заказчиком выступает кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов. Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель исследования - проверка четырех гипотез:**
1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

**Ход исследования**

Данные представлены заказчиком. О качестве данных ничего не известно, кроме описания столбцов. Таким образом, исследование пройдет в четыре этапа:
1. Обзор данных.
2. Предобработка данных.
3. Проверка гипотез.
4. Анализ результатов проверки и составление выводов.

## Обзор данных

In [1]:
# импортируем библиотеку pandas для работы с данными

import pandas as pd

# импортируем лемматизатор Mystem из библиотеки pymystem3

from pymystem3 import Mystem

# сохраним лемматизатор в переменной 'm', чтобы упростить использование в коде

m = Mystem() 

# импортируем Counter из библиотеки collections

from collections import Counter

# импортируем стеммер 

from nltk.stem import SnowballStemmer 

# читаем файл и сохраняем в переменную df дальнейшей работы, в том числе изменений
# используем оператор обработки исключений, чтобы избежать ошибки, при открытии данных,
# расположенных по другому адресу (мы храним данные в папке с тетрадью)

try:
    df = pd.read_csv('/datasets/data.csv')
except:
    df = pd.read_csv('data.csv')

# читаем файл и сохраняем в переменную df_control для контроля вносимых изменений
# сохраним переменную df_control неизменной

try:
    df_control = pd.read_csv('/datasets/data.csv')
except:
    df_control = pd.read_csv('data.csv')

# предварительно ознакомимся с данными, используем метод sample() для случайной выборки
# с параметром random_state=1, для сохранения результата при повторных запусках алгоритма

df_control.sample(10, random_state=1)

children  days_employed  dob_years education  education_id  \
1383          0  353802.811675         37   среднее             1   
300           1    -359.193975         33   СРЕДНЕЕ             1   
6565          2   -1064.854333         35   среднее             1   
17027         0            NaN         48    высшее             0   
4077          0   -7059.100220         45    высшее             0   
10437         0   -2258.301455         50   среднее             1   
9631          1   -1045.752744         23   среднее             1   
19867         0            NaN         39   среднее             1   
9124          0    -282.186487         48   среднее             1   
19634         0   -1320.103813         46   среднее             1   

          family_status  family_status_id gender income_type  debt  \
1383     вдовец / вдова                 2      F   пенсионер     0   
300    гражданский брак                 1      M   сотрудник     0   
6565   гражданский брак                 1      F   компаньон     0   
17027  гражданский брак                 1      F   сотрудник     0   
4077   гражданский брак                 1      F   компаньон     1   
10437   женат / замужем                 0      F   сотрудник     0   
9631    женат / замужем                 0      M   сотрудник     0   
19867   женат / замужем                 0      M   сотрудник     0   
9124          в разводе                 3      F   компаньон     0   
19634   женат / замужем                 0      F   компаньон     0   

        total_income                                purpose  
1383   216452.226085             строительство недвижимости  
300    223001.623994                  на проведение свадьбы  
6565   163591.209323                                свадьба  
17027            NaN                      операции с жильем  
4077   194820.185757                        сыграть свадьбу  
10437  146768.250599  операции с коммерческой недвижимостью  
9631    81734.568425  операции с коммерческой недвижимостью  
19867            NaN             дополнительное образование  
9124   284275.683334       строительство жилой недвижимости  
19634  184569.030590                             автомобили

In [2]:
# получим информацию о таблице
df_control.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов. Описание представлено заказчиком:
- children — количество детей в семье. Содержит целые числа.
- days_employed — общий трудовой стаж в днях. Содержит вещественные числа.
- dob_years — возраст клиента в годах. Содержит целые числа.
- education — уровень образования клиента. Содержит строки и другие данные, не подходящие ни под один тип (особенность типа данных object Pandas).
- education_id — идентификатор уровня образования. Содержит целые числа.
- family_status — семейное положение. Содержит строки и другие данные, не подходящие ни под один тип.
- family_status_id — идентификатор семейного положения. Содержит целые числа. 
[Словарь значений](#123)
- gender — пол клиента. Содержит строки и другие данные, не подходящие ни под один тип.
- income_type — тип занятости. Содержит строки и другие данные, не подходящие ни под один тип.
- debt — имел ли задолженность по возврату кредитов. Содержит целые числа.
- total_income — ежемесячный доход. Содержит вещественные числа.
- purpose — цель получения кредита. Содержит строки и другие данные, не подходящие ни под один тип.

Наименования столбцов соответствуют просмотренным данным, а также выполнены в хорошем стиле - их изменение не потребуется. 

Предварительное знакомство дало ряд наблюдений:
1. Количество значений в столбцах различается - имеются пропуски.
2. В столбце days_employed встречаются отрицательные значения - необходимо уточнить единицу измерения.
3. В столбце total_income указан ежемесячный доход - необходимо уточнить единицу измерения.

**Выводы**

Каждая строка содержит данные по клиенту банка. Предварительно можно утверждать, что данных достаточно для проверки гипотез. Однако в данных встречаются пропуски, а также дубликаты. Необходимо провести предобработку.

## Предобработка данных

### Стиль заголовков
Удостоверимся в хорошем стиле наименования колонок.

In [3]:
df_control.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

Пробелы отсутствуют, все записи сделаны в snake_case, используется понятный синтаксис. Как мы и предполагали - обработки наименования столбцов не потребуется.

### Пропуски значений
Посчитаем количество пропущенных значений в данных.

In [4]:
df_control.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
# подсчитаем количество дубликатов в данных 

df.duplicated().sum()

54

**Промежуточные выводы**
1. В колонках 'days_employed' и 'total_income' имеется одинаковое количество пропусков. Необходимо установить, связаны ли эти пропуски друг с другом, а также предположить причину их возникновения.


2. Общий трудовой стаж и его влияние на погашение кредита, не является предметом исследования, а значит эту колонку *можно было бы* удалить из данных. Но вместо этого, мы заполним ее средним значением, если в дальнейшем появится возможность уточнить с их помощью результаты исследования. При замене необходимо учитывать, что в столбце присутствуют отрицательные значения, а значит, нужно заменить текущие значения на абсолютные. 


3. Ежемесячный доход является необходимой частью исследования и пропуски в нем необходимо обработать, так как пропущенно около 10% всех значений, что может оказать существенное влияние на результат исследования. Пропущенные значения следует заменить медианным значением, так как большой разброс дохода может оказать существенное влияние на среднее значение.


4. В обоих столбцах содержатся количественные переменные. Пропуски вида NaN - используется при замещении отсутствующего числа и принадлежит к типу float, поэтому с ним можно проводить математические операции. Однако, пропуски необходимо заменить, в виду того, что они могут оказать влияние на результаты исследования как непосредственно (в случае с 'total_income'), так и опосредованно (при использовании 'days_employed' в дальнейшем уточнении результатов исследования).


5. В таблице имеется 54 явных дубликата.

#### Изучение столбцов с пропусками
Определим, связаны ли пропуски в столбцах 'days_employed' и 'total_income'.

In [6]:
# сравним два дата фрейма
# первый отфильтруем по пропускам в столбце 'days_employed', 
# второй отфильтруем по пропускам в столбце 'total_income'
# сравним их и суммируем результаты
# так как True принимается за 1, а False за 0, 
# итоговая сумма должна быть равна 2174, во всех столбцах, кроме сравниваемых

(df_control[df_control['days_employed'].isna()] == df_control[df_control['total_income'].isna()]).sum()

children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

Различия отсутствуют, а значит данные в указанных столбцах отсутствуют в одних и тех же строках. Разумно предположить, что при отсутствии сведений о трудовом стаже, не указывались и сведения о доходе заемщика. 

Посмотрим на максимальное и миниальное значение столбцов.

In [7]:
# максимальное и минимальное значения 'days_employed'

print('Максимальное значение трудового стажа:', 
      int(df_control['days_employed'].max()))
print('Минимальное значение трудового стажа:', 
      int(df_control['days_employed'].min()))

# максимальное и минимальное значения 'total_income'

print('Максимальное значение ежемесячного дохода:', 
      int(df_control['total_income'].max()))
print('Минимальное значение ежемесячного дохода:', 
      int(df_control['total_income'].min()))

Максимальное значение трудового стажа: 401755
Минимальное значение трудового стажа: -18388
Максимальное значение ежемесячного дохода: 2265604
Минимальное значение ежемесячного дохода: 20667


**Предварительные выводы:**
1. Максимальный трудовой стаж составляет 401 тысячу дней. Явная ошибка при формировании данных в этом столбце. В настоящий момент, уточнить или изменить эти данные можно только основываясь на предположениях. Если в дальнейшем потребуется обращаться к этим данным, необходимо будет уточнить единицу измерения и возможные ошибки. Есть гипотеза, что запись производилась в днях * 24. **В любом случае, стоит обратить внимание разработчиков на этот артефакт.**
2. Минимальный трудовой стаж имеет отрицательное значение, что может повлиять на рассчеты с этим столбцом.
3. Значения ежемесячного дохода не вызывают вопросов.

#### Обработка столбца 'days_employed'
Для условий задания отсутствует необходимость обработки этих данных, но в рамках тренировки, хочется обработать эту колонку, как одну из самых проблемных.

In [8]:
# предположим, что можно обеспечить более точное заполнение пропусков
# в столбце 'days_employed', если выявим связь с другими значениями
#  для этого посчитаем коэффициент корреляции между данными, 
# содержащимися в столбцах 'days_employed' и 'dob_years'

df['days_employed'].corr(df['dob_years'], "pearson")

0.5826429853905242

r = 0.58 
Не лучшее значение, чтобы увеличить его, попробуем добавить новую классификацию по возрасту, выделив различные возрастные группы, после чего посчитать r уже на обновленных данных.
Сперва нужно определиться с возрастной периодизацией. В современной науке этот параметр распространен достаточно широко. Однако, с учетом социально-экономических особенностей устройства российского общества, использовать мировые классификации или классификацию ВОЗ нам представляется неразумным. К сожалению, в настоящее время, в российской науке отсутствует единая стандартизация в вопросах возрастной периодизации, поэтому мы сошлемся на следующее, до сих пор популярное, решение.


**Возрастная периодизаций 1965 года на симпозиуме Академии педагогических наук СССР:**

Данная периодизация представляется нам наиболее подходящей для представленной задачи, учитывая постсоветское наследие на территории Российской Федерации.

    - Юношеский возраст — от 17 до 21 года (муж.); от 16 до 20 лет (жен.)
    - Средний возраст
    - первый период — от 22 до 35 года (муж.); от 21 до 35 лет (жен.)
    - второй период — от 36 до 60 года (муж.); от 36 до 55 лет (жен.)
    - Пожилые люди — от 61 до 75 года (муж.); от 56 до 75 лет (жен.)

Объединим представленные выше группы в возрастные классы, разделив категорию 36-60 на две, соответствующих классификации ВОЗ:


- 1: до 21 года
- 2: 22-35 лет
- 3: 36-45 лет 
- 4: 46-60 лет
- 5: 61+

Напишем подходящую функцию, после чего добавим полученные данные в таблицу и пересчитаем r.

In [9]:
# напишем функцию
def age_classifier(dob_years, age_class_1=21, 
                   age_class_2=35, age_class_3=45, age_class_4=60):
    '''
    Функция age_classifier принимает в параметры 'dob_years'
    возраст заемщика, после чего возвращает 
    соответствующее ему значение возрастной периодизации, в зависимости 
    от величины приданной функции сумме в параметрах age_class_*. 
    
    В качестве значения по умолчанию параметров age_class_*,
    используется выбранная нами модель возрастной периодизации,
    основанная на симпозиуме АПН 1965 года.
    
    '''
    if dob_years <= age_class_1:
        return 1
    if dob_years <= age_class_2:
        return 2
    if dob_years <= age_class_3:
        return 3
    if dob_years <= age_class_4:
        return 4
    return 5

# протестируем функцию для каждой группы
print(age_classifier(19))
print(age_classifier(27))
print(age_classifier(45))
print(age_classifier(55))
print(age_classifier(67))

1
2
3
4
5


In [10]:
# добавим столбец с возрастной периодизацией в таблицу

df['age_class'] = df['dob_years'].apply(age_classifier)

# проверим результат работы функции

display(df.head())

# оценим получившийся разброс
df['age_class'].value_counts()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  age_class  
0               покупка жилья          3  
1     приобретение автомобиля          3  
2               покупка жилья          2  
3  дополнительное образование          2  
4             сыграть свадьбу          4

4    7057
2    6418
3    5633
5    2140
1     277
Name: age_class, dtype: int64

Отличное статистическое распределение, за исключением первой группы. Но её выделение обосновано экономически и законодательно, поэтому оставим ее в составе данных. Перейдем к рассчету коэффициента корреляции.

In [11]:
df['days_employed'].corr(df['age_class'], "pearson")

0.5434754563994446

Коэффициент уменьшился на 4%. Можно сделать вывод, что возрастная периодизация не имеет связи с трудовым стажем в данных, что безусловно связано с ошибкой заполнения столбца 'days_employed'. Сгруппируем среднее значение стажа по возрастным группам.

In [12]:
# создадим сводные таблицы, вызвав функции для изучения цифр
display(df.pivot_table(index=['age_class'], values='days_employed', aggfunc=['max', 'min','mean', 'median']))

max           min           mean         median
           days_employed days_employed  days_employed  days_employed
age_class                                                           
1          400992.375704 -10689.250498   24257.558395    -757.612624
2          397905.003200  -7622.710756    -500.022275   -1224.714188
3          400553.798436 -11037.198423    2725.927168   -1834.869762
4          401755.400475 -16593.472817  102779.971304   -1185.248658
5          401715.811749 -18388.949901  289109.453187  356191.137667

А вот и ошибка в данных. Данные искажены из-за крайних значений. При этом, максимальные значения для 4-5 возрастной группы выглядят легитимно, если переводить из дней в часы. И абсолютно чуждо для групп с 1 по 3. 


Максимальное значение, если оценивать его в часах будет равняться приблизительно 45 годам. Что похоже на усредненное значение по заемщикам. Косвенно этот факт подтверждается тем, что 4 возрастная группа (46-60 лет - возрастное попадание в стаж пракитчески идеально!), самая широкопредставленная в возрастном массиве. 


Из этого можно сделать вывод, что данные по стажу первично содержали большое количество пропусков, а после обработки, пропуски были заполнены средним значением по всем заемщикам. При этом, по какой-то причине, замена проихводилась в единицах (дни стажа * 24). Что и привело к искажению среднего стажа всех возрасных групп. К тому же нельзя забывать об очевидной ошибке в виде отрицательного значения стажа.

Логично предположить, что замену стоит произвести по прогнозируемому среднему стажу в возрастной группе, исходя из занятости с 18 лет, при этом нужно не обрабатывать пропуски, а провести сквозную полную замену из-за крайне низкого качества данных.

In [13]:
# напишем функцию, рассчитывающую средний стаж, исходя из возрастной группы
# и законодательного возраста начала рабочих отношений 18 лет

def days_employed_oracle(dob_years, age_class_1=21, 
                   age_class_2=35, age_class_3=45, age_class_4=60):
    '''
    Функция days_employed_oracle принимает в параметры 'dob_years'
    возраст заемщика, после чего возвращает 
    соответствующее ему значение среднего трудового стажа
    от величины приданной функции сумме в параметрах age_class_*. 
    
    В качестве значения по умолчанию параметров age_class_*,
    используется выбранная нами модель возрастной периодизации,
    основанная на симпозиуме АПН 1965 года.
    
    Вычитание 18 обеспечивает отсутствие учета первых 18 лет
    жизни в качестве рабочего стажа.
    Умножение на 365 позволяет привести значение к заданным
    технической документацией параметрам (измерение стажа в днях).
    
    '''
    if dob_years <= age_class_1:
        return (age_class_1 - 18) / 2 * 365
    if dob_years <= age_class_2:
        return (age_class_2 + age_class_1 - 18) / 2 * 365
    if dob_years <= age_class_3:
        return (age_class_3 + age_class_2 - 18) / 2 * 365
    if dob_years <= age_class_4:
        return (age_class_4 + age_class_3 - 18) / 2 * 365
    return (65 + age_class_4 - 18) / 2 *365

# протестируем функцию для каждой группы
print(days_employed_oracle(19))
print(days_employed_oracle(27))
print(days_employed_oracle(45))
print(days_employed_oracle(55))
print(days_employed_oracle(67))

547.5
6935.0
11315.0
15877.5
19527.5


In [14]:
# заменим пропуски в колонке 'days_employed' на среднее значение
# трудового стажа для возрастных групп

df['days_employed'] = df['dob_years'].apply(days_employed_oracle).astype('int')

# проверим результат работы, запросив количество пропусков, 
# а также изменение среднего значения столбца

mean_days_employed_before = df_control['days_employed'].abs().mean()
mean_days_employed_after = df['days_employed'].mean()

print('Количество пропусков в столбце \'days_employed\':', df['days_employed'].isna().sum())
print('Среднее значение трудового стажа в столбце \'days_employed\' до замены: ' 
      + str(round(mean_days_employed_before / 365))  + ' года.')
print('Среднее значение трудового стажа в столбце \'days_employed\' после замены: ' 
      + str(round(mean_days_employed_after / 365)) + ' года.')
print(f'Среднее значение столбца \'days_employed\' изменилось ' 
      f'на {mean_days_employed_before / mean_days_employed_after:.1%}')

df.sample(5, random_state=1)

Количество пропусков в столбце 'days_employed': 0
Среднее значение трудового стажа в столбце 'days_employed' до замены: 183 года.
Среднее значение трудового стажа в столбце 'days_employed' после замены: 33 года.
Среднее значение столбца 'days_employed' изменилось на 549.3%


children  days_employed  dob_years education  education_id  \
1383          0          11315         37   среднее             1   
300           1           6935         33   СРЕДНЕЕ             1   
6565          2           6935         35   среднее             1   
17027         0          15877         48    высшее             0   
4077          0          11315         45    высшее             0   

          family_status  family_status_id gender income_type  debt  \
1383     вдовец / вдова                 2      F   пенсионер     0   
300    гражданский брак                 1      M   сотрудник     0   
6565   гражданский брак                 1      F   компаньон     0   
17027  гражданский брак                 1      F   сотрудник     0   
4077   гражданский брак                 1      F   компаньон     1   

        total_income                     purpose  age_class  
1383   216452.226085  строительство недвижимости          3  
300    223001.623994       на проведение свадьбы          2  
6565   163591.209323                     свадьба          2  
17027            NaN           операции с жильем          4  
4077   194820.185757             сыграть свадьбу          3

549%! Почти в шесть раз была допущена ошибка при формировании данных в этом столбце. Немудрено, ведь средний стаж заемщика составлял 183 года. Всем бы таких работников.

Пропуски полностью обработаны и замененые средним значением по столбцу. 
Среднее значение столбца 'days_employed' изменилось на 549%, что можно считать существенным вкладом в увеличение точности исследования.

#### Обработка столбца 'total_income'

Для заполнения столбцов можно использовать разные способы. Но сперва проверим гипотезу, о зависимости дохода от типа занятости.

In [15]:
# создадим сводные таблицы, вызвав функции для изучения цифр

# проверим качество данных перед группировкой

print(df['income_type'].unique())

# построим сводную таблицу, содержащую сведения о доходах заемщиков, 
# исходя из их типа дохода

display(df.pivot_table(index=['income_type'], values='total_income', aggfunc=['max', 'min', 'mean', 'median']))

['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']


max            min           mean         median
                 total_income   total_income   total_income   total_income
income_type                                                               
безработный      2.027225e+05   59956.991984  131339.751676  131339.751676
в декрете        5.382913e+04   53829.130729   53829.130729   53829.130729
госслужащий      9.104515e+05   29200.077193  170898.309923  150447.935283
компаньон        2.265604e+06   28702.812889  202417.461462  172357.950966
пенсионер        7.351033e+05   20667.263793  137127.465690  118514.486412
предприниматель  4.991631e+05  499163.144947  499163.144947  499163.144947
сотрудник        1.726276e+06   21367.648356  161380.260488  142594.396847
студент          9.820163e+04   98201.625314   98201.625314   98201.625314

Выведенные значения вызывают некоторые сомнения. Складывается ощущение, что для типа занятости "в декрете", а также "студент" были введены некоторые условные данные. Студент со средним доходом 98 тысяч рублей - мечты, мечты. При этом, учитывая специфику социально-экономических условий в Российской Федерации, доход безработного в 131 тысячу вполне может оказаться реальностью. В целом, опираясь на полученные цифры, значения дохода выглядят завышенными. 


Есть вероятность применения каких-то внутренних банковских документов, либо банальная ошибка. В любом случае, изменять доход, основываясь на экономических исследованиях, мы не будем. Проведем простую заменую пропущенных значений, исходя из вида занятости клиента.

In [16]:
# проведем замену пропущенных значений, 
# исходя из группировки по виду деятельности

df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform('median'))

# вывод количества пропусков до и после обработки

print('Количество пропусков в столбце \'total_income\' до замены:', 
      df_control['total_income'].isna().sum())
print('Количество пропусков в столбце \'total_income\' после замены:', 
      df['total_income'].isna().sum())

# вывод средних значений до и после обработки 
# и их изменение
median_total_income_before = round(df_control['total_income'].median(), 2)
median_total_income_after = round(df['total_income'].median(), 2)

print(f'Медианное значение в столбце \'total_income\' до замены: ' 
      f'{median_total_income_before:.1f}')
print(f'Медианное значение в столбце \'total_income\' после замены: ' 
      f'{median_total_income_after:.1f}')

print(f'Медианное значение столбца \'total_income\' изменилось ' 
      f'на {median_total_income_after / median_total_income_before -1:.1%}')

# вывод образца таблицы, для визуального контроля изменений
display(df.sample(5, random_state=1))

Количество пропусков в столбце 'total_income' до замены: 2174
Количество пропусков в столбце 'total_income' после замены: 0
Медианное значение в столбце 'total_income' до замены: 145017.9
Медианное значение в столбце 'total_income' после замены: 142594.4
Медианное значение столбца 'total_income' изменилось на -1.7%


children  days_employed  dob_years education  education_id  \
1383          0          11315         37   среднее             1   
300           1           6935         33   СРЕДНЕЕ             1   
6565          2           6935         35   среднее             1   
17027         0          15877         48    высшее             0   
4077          0          11315         45    высшее             0   

          family_status  family_status_id gender income_type  debt  \
1383     вдовец / вдова                 2      F   пенсионер     0   
300    гражданский брак                 1      M   сотрудник     0   
6565   гражданский брак                 1      F   компаньон     0   
17027  гражданский брак                 1      F   сотрудник     0   
4077   гражданский брак                 1      F   компаньон     1   

        total_income                     purpose  age_class  
1383   216452.226085  строительство недвижимости          3  
300    223001.623994       на проведение свадьбы          2  
6565   163591.209323                     свадьба          2  
17027  142594.396847           операции с жильем          4  
4077   194820.185757             сыграть свадьбу          3

In [17]:
# максимальное и минимальное значения 'total_income'

print('Максимальное значение ежемесячного дохода:', 
      int(df_control['total_income'].max()))
print('Минимальное значение ежемесячного дохода:', 
      int(df_control['total_income'].min()))

Максимальное значение ежемесячного дохода: 2265604
Минимальное значение ежемесячного дохода: 20667


Разброс цифр слишком большой, необходимо найти экономические основания для классификации данных.


Согласно статистике Росстата за 2020 года медианная зарплата россиян составила 32422 рубля, средняя — 51352 рубля.
Необходимо уточнить, что на указанные цифры оказывает сильное влияние регион проживания. 


Однако, в представленных данных регион отсутствует, поэтому будем производить классификацию по медианной зарплате.
Сперва разобьем на группы кратные медианной зарплате и оценим их количество.

In [18]:
# оценим количество групп при таком подходе
int(df_control['total_income'].max()) / 32422

69.87860094997224

70 групп - хорошо для статистики, плохо для анализа. Надо изменять параметр классификации данных. Статистически верным решением, будет классификация по медианному значению этого столбца. Проверим теорию.

In [19]:
int(df_control['total_income'].max()) / median_total_income_after

15.888450037308619

16 групп. Уже лучше, но все еще слишком высокая детализация. Предложим следующую классификацию, основываясь на нелинейном росте зарплат:
1. Класс 0: 1/2 и менее медианной зарплаты по данным.
2. Класс 1: 2/3 и менее медианной зарплаты по данным.
3. Класс 2: медианная зарплата по данным и менее.
4. Класс 3: полторы и менее медианных зарплаты по данным. 
5. Класс 4: две и менее медианных зарплат по данным.
6. Класс 5: свыше двух медианных зарплат по данным.

Проведем классификацию, после чего оценим получившиеся объемы данных.

In [20]:
# напишем функцию
def income_classifier(total_income, base=median_total_income_after):
    '''
    Функция income_classifier принимает в параметры 'total_income'
    числовое значение дохода, после чего возвращает 
    соответствующее ему значение классификации, в зависимости 
    от величины приданной функции сумме в параметре base. 
    
    В качестве значения по умолчанию параметра base
    выбрана переменная median_total_income_after,
    содержащая данные о медианном значении df['total_income']
    после заполнения пропусков.
    '''
    if total_income <= base / 2:
        return 0
    if total_income <= base / 3 * 2:
        return 1
    if total_income <= base:
        return 2
    if total_income <= base * 1.5:
        return 3
    if total_income <= base * 2:
        return 4
    return 5

# протестируем функцию для каждой группы
print(income_classifier(50000))
print(income_classifier(90000))
print(income_classifier(140000))
print(income_classifier(170000))
print(income_classifier(300000))
print(income_classifier(500000))

0
1
2
3
5
5


In [21]:
# добавим столбец с категоризацией в таблицу
# для чего введем тестовую переменную, чтобы провести эксперименты

df_value = df
df_value['income_class'] = df_value['total_income'].apply(income_classifier)

# проверим результат работы функции

display(df_value.head())

# оценим получившийся разброс
df_value['income_class'].value_counts()

children  days_employed  dob_years education  education_id  \
0         1          11315         42    высшее             0   
1         1          11315         36   среднее             1   
2         0           6935         33   Среднее             1   
3         3           6935         32   среднее             1   
4         0          15877         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  age_class  income_class  
0               покупка жилья          3             4  
1     приобретение автомобиля          3             2  
2               покупка жилья          2             3  
3  дополнительное образование          2             4  
4             сыграть свадьбу          4             3

2    7006
3    6300
4    2563
1    2320
5    1757
0    1579
Name: income_class, dtype: int64

Как видно из результатов мы добились неплохой репрезентативности выборки, учитывая большой разброс доходов заемщиков. Эти данные можно использовать в дальнейшем для исследования, а также дополнить ими изучаемые данные в переменной df.

In [22]:
df['income_class'] = df_value['income_class']
df.sample(10, random_state=1)

children  days_employed  dob_years education  education_id  \
1383          0          11315         37   среднее             1   
300           1           6935         33   СРЕДНЕЕ             1   
6565          2           6935         35   среднее             1   
17027         0          15877         48    высшее             0   
4077          0          11315         45    высшее             0   
10437         0          15877         50   среднее             1   
9631          1           6935         23   среднее             1   
19867         0          11315         39   среднее             1   
9124          0          15877         48   среднее             1   
19634         0          15877         46   среднее             1   

          family_status  family_status_id gender income_type  debt  \
1383     вдовец / вдова                 2      F   пенсионер     0   
300    гражданский брак                 1      M   сотрудник     0   
6565   гражданский брак                 1      F   компаньон     0   
17027  гражданский брак                 1      F   сотрудник     0   
4077   гражданский брак                 1      F   компаньон     1   
10437   женат / замужем                 0      F   сотрудник     0   
9631    женат / замужем                 0      M   сотрудник     0   
19867   женат / замужем                 0      M   сотрудник     0   
9124          в разводе                 3      F   компаньон     0   
19634   женат / замужем                 0      F   компаньон     0   

        total_income                                purpose  age_class  \
1383   216452.226085             строительство недвижимости          3   
300    223001.623994                  на проведение свадьбы          2   
6565   163591.209323                                свадьба          2   
17027  142594.396847                      операции с жильем          4   
4077   194820.185757                        сыграть свадьбу          3   
10437  146768.250599  операции с коммерческой недвижимостью          4   
9631    81734.568425  операции с коммерческой недвижимостью          2   
19867  142594.396847             дополнительное образование          3   
9124   284275.683334       строительство жилой недвижимости          4   
19634  184569.030590                             автомобили          4   

       income_class  
1383              4  
300               4  
6565              3  
17027             2  
4077              3  
10437             3  
9631              1  
19867             2  
9124              4  
19634             3

Проверим данные на предмет пропусков, перед переходом к следующему этапу исследования.

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  int32  
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
 12  age_class         21525 non-null  int64  
 13  income_class      21525 non-null  int64  
dtypes: float64(1), int32(1), int64(7), object(5)
memory usage: 2.2+ MB


### Обработка дубликатов

#### Неявные дубликаты
Перейдем к поиску неявных дубликатов. Из обзора данных ясно, что основная проблема может заключаться в единообразии текстовых полей. Поэтому последовательно проверим столбцы со строками на уникальные значения, приведя их к единому виду.

In [24]:
# изучим данные по количеству детей

df['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

Разброс достаточно странный, вряд ли имеются в виду 20 или -1 ребенок. Посмотрим на частоту появления значений.

In [25]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Частота артефактных значений (-1 и 20) невелика, а значит можно удалить их или заменить на медианное значение. В связи с небольшим количеством значений, это не должно повлиять на чистоту исследования.

In [26]:
# удалим выбросы
df = df[df['children'] != 20]
df = df[df['children'] != -1]

# # проверим результат
df['children'].value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

In [27]:
df['education'].unique()

# приведем их значения к одному регистру

df['education'] = df['education'].str.lower()

# проверим результат

df['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

In [28]:
df['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [29]:
df['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [30]:
df['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [31]:
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Промежуточные выводы**
1. Несмотря на определенное влияние приведенных выше столбцов на возврат кредита, от клиента не поступало запрос на анализ указанных метрик. В связи с чем, в виду большого объема работы по приведению в соответствие, столбцы, содержащие дополнительную информацию, стоит удалить.
2. Кроме того, установить соответствие между столбцами 'family_status' и 'family_status_id'.
3. Столбец 'purpose', необходимый для исследования, содержит большое количество дубликатов и нарушения стиля.

#### Явные дубликаты
Проверим данные на явные дубликаты, учитывая произведенные нами замены.

In [32]:
# проверим данные на явные дубликаты

df.duplicated().sum()

71

**Промежуточные выводы**


Учитывая отсутствие уникального номера клиента, мы можем предположить, что данные сведения являются ошибкой. Учитывая их небольшое количество (71) в отношении к общему количеству записей (21525), их можно отбросить и это не повлияет на точность исследования. Если в дальнейшем получится установить уникальность этих записей (например, id клиента), то эти дубликаты можно будет восстановить и использовать в повторном анализе.


Избавляемся от явных дубликатов.

In [33]:
# применим метод reset_index(), чтобы сбросить индес таблицы
# используем параметр drop=True, чтобы не создавать столбец со старыми индексами

df = df.drop_duplicates().reset_index(drop=True)

# проверим себя и подсчитаем число дубликатов после удаления

df.duplicated().sum()

0

##### Обработка столбцов 'family_status' и 'family_status_id'
Для исследования нам хватает данных столбца 'family_status'. Но в технической документации, переданной заказчиком, отсутствует расшифровка значений 'family_status_id', поэтому создадим таблицу-словарь и прикрепим его ссылкой на ячейку к описанию документации, на случай необходимости в дальнейшем.

In [34]:
# так как мы будем использовать в дальнейшем значения столбца family_status
# приведем их к единому виду

df['family_status'] = df['family_status'].str.lower()

# создадим отдельную таблицу-словарь значений 'family_status'

family_status_dict = df[['family_status', 'family_status_id']]

# посмотрим на количество дубликатов
print(family_status_dict.duplicated().sum())

# проверим проведенную работу
family_status_dict.head()

21326


family_status  family_status_id
0   женат / замужем                 0
1   женат / замужем                 0
2   женат / замужем                 0
3   женат / замужем                 0
4  гражданский брак                 1

Большое количество дубликатов необходимо подчистить для создания таблицы-словаря.

In [35]:
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)

<a id="123"></a>
Дубликаты подчищены и словарь соответствия готов. 

In [36]:
family_status_dict

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

##### Обработка столбца 'purpose'
Произведем замену дубликатов в колонце 'purpose'. Для этого сперва изучим подробнее состав столбца.

In [37]:
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

В столбце содержится большое число неявных дубликатов, установить которые не имеется возможности. Например, есть 'операции с коммерческой недвижимостью', а также 'операции со своей недвижимостью'. И, хотя, операции с жилой и коммерческой недвижимостью стоит разделить на две группы, данные не содержат информации о том, как это можно сделать корректно. Единственным критерием в определении цели кредита в представленных данных будет предмет сделки. 

Чтобы изучить подрбонее столбец, посмотрим как часто встречаются какие слова.

In [38]:
words = df['purpose'].unique()

In [39]:
# выделим леммы из столбца

lemma_list = []
for word in words:
    lemma_word = m.lemmatize(word)
    lemma_list += lemma_word

# проверим результат

lemma_list

['покупка',
 ' ',
 'жилье',
 '\n',
 'приобретение',
 ' ',
 'автомобиль',
 '\n',
 'дополнительный',
 ' ',
 'образование',
 '\n',
 'сыграть',
 ' ',
 'свадьба',
 '\n',
 'операция',
 ' ',
 'с',
 ' ',
 'жилье',
 '\n',
 'образование',
 '\n',
 'на',
 ' ',
 'проведение',
 ' ',
 'свадьба',
 '\n',
 'покупка',
 ' ',
 'жилье',
 ' ',
 'для',
 ' ',
 'семья',
 '\n',
 'покупка',
 ' ',
 'недвижимость',
 '\n',
 'покупка',
 ' ',
 'коммерческий',
 ' ',
 'недвижимость',
 '\n',
 'покупка',
 ' ',
 'жилой',
 ' ',
 'недвижимость',
 '\n',
 'строительство',
 ' ',
 'собственный',
 ' ',
 'недвижимость',
 '\n',
 'недвижимость',
 '\n',
 'строительство',
 ' ',
 'недвижимость',
 '\n',
 'на',
 ' ',
 'покупка',
 ' ',
 'подержать',
 ' ',
 'автомобиль',
 '\n',
 'на',
 ' ',
 'покупка',
 ' ',
 'свой',
 ' ',
 'автомобиль',
 '\n',
 'операция',
 ' ',
 'с',
 ' ',
 'коммерческий',
 ' ',
 'недвижимость',
 '\n',
 'строительство',
 ' ',
 'жилой',
 ' ',
 'недвижимость',
 '\n',
 'жилье',
 '\n',
 'операция',
 ' ',
 'со',
 ' ',
 'свой'

In [40]:
# # подсчитаем частоту встречаемых слов
Counter(lemma_list)

Counter({'покупка': 10,
         ' ': 59,
         'жилье': 7,
         '\n': 38,
         'приобретение': 1,
         'автомобиль': 9,
         'дополнительный': 2,
         'образование': 9,
         'сыграть': 1,
         'свадьба': 3,
         'операция': 4,
         'с': 5,
         'на': 4,
         'проведение': 1,
         'для': 2,
         'семья': 1,
         'недвижимость': 10,
         'коммерческий': 2,
         'жилой': 2,
         'строительство': 3,
         'собственный': 1,
         'подержать': 1,
         'свой': 4,
         'со': 1,
         'заниматься': 2,
         'сделка': 2,
         'подержанный': 1,
         'получение': 3,
         'высокий': 3,
         'профильный': 1,
         'сдача': 1,
         'ремонт': 1})

Выделим, в качестве гипотезы, основные слова, описывающие сделки: 'недвижимость', 'автомобиль', 'образование', 'жилье', 'свадьба', 'жилой'

Чтобы объединить значения по ним, выведем из них словарные основы и посмотрим их частоту в тексте.

In [41]:
# соберем слова в список

credit_purpose = ['недвижимость', 'автомобиль', 'образование', 'жилье', 'свадьба', 'жилой']

# найдем основу этих слов
# передаем стеммеру значение 'russian', чтобы включить 
# лингвистический анализ русского языка

russian_stemmer = SnowballStemmer('russian')

stem_credit_purpose = []
for word in credit_purpose:
    print ('Исходное слово - ' + word + ', после стемминга - ' + russian_stemmer.stem(word)) 
    stem_credit_purpose.append(russian_stemmer.stem(word))
print()
print('Основа слова, определяющего предмет кредита: ', stem_credit_purpose)

Исходное слово - недвижимость, после стемминга - недвижим
Исходное слово - автомобиль, после стемминга - автомобил
Исходное слово - образование, после стемминга - образован
Исходное слово - жилье, после стемминга - жил
Исходное слово - свадьба, после стемминга - свадьб
Исходное слово - жилой, после стемминга - жил

Основа слова, определяющего предмет кредита:  ['недвижим', 'автомобил', 'образован', 'жил', 'свадьб', 'жил']


Исходя из полученного списка, можно сделать вывод, что два слова дали одну основу - 'жил'. Это не противоречит полученным данным. Теперь изучим репрезентативность основ в полном списке целей кредита.

In [42]:
# удалим лишнее значение из списка, определяющего предмет кредита
stem_credit_purpose.pop()
print('Окончательный список основ слов, ' 
      'определяющих предмет кредита: ', (stem_credit_purpose))

# для этого посчитаем их контейнером Counter
Counter(russian_stemmer.stem(word) for word in lemma_list)

Окончательный список основ слов, определяющих предмет кредита:  ['недвижим', 'автомобил', 'образован', 'жил', 'свадьб']


Counter({'покупк': 10,
         ' ': 59,
         'жил': 9,
         '\n': 38,
         'приобретен': 1,
         'автомобил': 9,
         'дополнительн': 2,
         'образован': 9,
         'сыгра': 1,
         'свадьб': 3,
         'операц': 4,
         'с': 5,
         'на': 4,
         'проведен': 1,
         'для': 2,
         'сем': 1,
         'недвижим': 10,
         'коммерческ': 2,
         'строительств': 3,
         'собствен': 1,
         'подержа': 2,
         'сво': 4,
         'со': 1,
         'занима': 2,
         'сделк': 2,
         'получен': 3,
         'высок': 3,
         'профильн': 1,
         'сдач': 1,
         'ремонт': 1})

**Промежуточные выводы:**
1. Из представленного текстового массива можно выделить основные предметы сделок, это:
    - сделки с недвижимостью
    - сделки с автомобилями
    - оплата образования 
    - проведение свадьбы
    
    Рассмотрим последовательно каждый вид сделок, чтобы оценить точность информации, полученной из массива.


2. Сделки с недвижимостью принято подразделять по предмету: на сделки с коммерческой и жилой, а также виду: аренда, купля, ремонт и прочее. Из представленных данных можно сделать следующие выводы: "жил" встречается 9 раз, "недвижим" - 10, а "коммерческ" - 2. Из этих данных отсутствует возможность сделать вывод о конкретном предмете сделки с недвижимостью - коммерческой или жилой. Кроме того, пропуски в описании вида сделки также невозможно восстановить. **Вывод:** это удручает, так как подобное разделение может оказать большое влияние на точность исследования. В любом случае, об этом стоит сообщить заказчику. А для продолжения исследования объединить жилую и коммерческую недвижимость в одну группу "сделки с недвижимостью".


3. Сделки с автомобилями можно разделить также на сделки для коммерческой или личной цели - сведения полностью отсутствуют. И по виду - представленные данные и здесь не радуют своим искажением. **Вывод**: что ж, сообщим заказчику о слабости этой метрики и объединим сделки в "сделки с автомобилями.


4. Сделки, связанные с получением и оплатой образования. В данных есть вся информация, иходя из которой можно определить влияние вида образования на возврат кредита. А значит цель кредита на взятие какого-либо образования (например, дополнительного или высшего) вполне может коррелироваться с оценкой заемщика. Это вполне может стать интересным дополнительным кейсом, который окажет влияние на общие сведения. Но в настоящее время, это субисследование не требуется. **Вывод:** значит можно объединить все сделки в "получение образования".


5. Кредиты на проведение свадьбы не требуют дополнительных пояснений и данные по ним могут быть использованы as is.

**Итоги:** выделим следующие ключевые слова "недвижимость", "автомобиль", "образование", "жилье", "свадьба". Проведем стемминг, для выяснения основы слова (чтобы осуществить замену в таблице), а также заменим все сущности "жилья" на "недвижимость.

Перейдем к обработке столбца 'purpose'.

In [43]:
# напишем функцию, для возврата основы слова из 
# списка значений в целевом столбце

def stemmator(column, base=stem_credit_purpose, exception='жил', exception_return='недвижим'):
    '''
    Функция stemmator проверяет значения в строке 
    столбца, выбранного параметром column. 
    Если это значение есть в списке параметра base, 
    то фукнция возвращает это значение из списка параметра base.
    
    Функции можно придать дополнительные параметры:
    exception - исключение из параметра base, позволяющее 
    возвратить уникальное значение исключения, зафиксированное 
    в параметре exception_return.
    
    Текущие значения по умолчанию параметров exception и exception_return
    обусловлены необходимостью объединить сущности 'жил' и 'недвижим'
    
    '''
    for row in column:
        for word in base:
            if word == exception:
                if word in row:
                    return exception_return
            if word in row:
                return word

# обновим данные столбца 'purpose', разбив значения,
# чтобы функция обработки могла пройтись по всем значениям в строке
# после чего применим функцию

df['purpose_class'] = df['purpose'].str.split().apply(stemmator)

# проверим результат работы функции

df.sample(10, random_state=1)

children  days_employed  dob_years education  education_id  \
18453         0          15877         48   среднее             1   
4398          1          11315         36   среднее             1   
18147         0          15877         57   среднее             1   
1305          1           6935         29   среднее             1   
625           1           6935         34    высшее             0   
1073          0          11315         43   среднее             1   
21115         2          11315         36    высшее             0   
16921         1           6935         32   среднее             1   
19274         0          19527         65   среднее             1   
13023         1           6935         34    высшее             0   

          family_status  family_status_id gender income_type  debt  \
18453   женат / замужем                 0      F   сотрудник     1   
4398          в разводе                 3      M   сотрудник     0   
18147   женат / замужем                 0      F   пенсионер     0   
1305    женат / замужем                 0      F   компаньон     0   
625    гражданский брак                 1      M   сотрудник     0   
1073    женат / замужем                 0      M   компаньон     0   
21115  гражданский брак                 1      F   компаньон     0   
16921  гражданский брак                 1      F   сотрудник     0   
19274   женат / замужем                 0      F   пенсионер     0   
13023   женат / замужем                 0      F   компаньон     0   

        total_income                                 purpose  age_class  \
18453   71793.177187         операции со своей недвижимостью          4   
4398   110795.035534            на покупку своего автомобиля          3   
18147  146046.360232                       операции с жильем          4   
1305   172357.950966                   получение образования          2   
625    142594.396847                         сыграть свадьбу          2   
1073   124344.964970   получение дополнительного образования          3   
21115  327135.268790                                 свадьба          3   
16921  211255.376700                                 свадьба          2   
19274   52352.253479                            недвижимость          5   
13023  186841.217645  строительство собственной недвижимости          2   

       income_class purpose_class  
18453             1      недвижим  
4398              2     автомобил  
18147             3      недвижим  
1305              3     образован  
625               2        свадьб  
1073              2     образован  
21115             5        свадьб  
16921             3        свадьб  
19274             0      недвижим  
13023             3      недвижим

**Промежуточные выводы:**
Все данные подготовлены для исследования, удалим лишние данные, после чего перейдем к изучению гипотез.

##### Удаление лишних данных
Удалим лишние для исследования столбцы, чтобы облегчить работу с данными.

In [44]:
# создадим новую переменную, чтобы использовать её в исследовании

df_clear = df.drop(columns=[
    'days_employed', 'dob_years', 'education', 
    'education_id', 'family_status_id', 
    'gender', 'income_type', 'total_income',
    'purpose'
])
df_clear.sample(5, random_state=1)

children     family_status  debt  age_class  income_class purpose_class
18453         0   женат / замужем     1          4             1      недвижим
4398          1         в разводе     0          3             2     автомобил
18147         0   женат / замужем     0          4             3      недвижим
1305          1   женат / замужем     0          2             3     образован
625           1  гражданский брак     0          2             2        свадьб

## Проведение исследования
### Изучение статистики
Заказчиком были поставлены следующие цели исследования:
1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

Перед началом исследования изучим статистику.

In [45]:
debt_stat = df_clear['debt'].value_counts()

print('Имеют задолженность по возврату кредита у ' 
      + str(debt_stat[0]) + ' заемщиков.')

print('Задолженность по кредиту отсутствует у ' 
      + str(debt_stat[1]) + ' заемщиков.')

print(f'Общий процент возвратов без просрочки: {debt_stat[1] / debt_stat[0]:.7%}')

Имеют задолженность по возврату кредита у 19599 заемщиков.
Задолженность по кредиту отсутствует у 1732 заемщиков.
Общий процент возвратов без просрочки: 8.8371856%


Данные неутешительные. Но дают нам представление о том, что больших цифр в исследовании ждать не стоит. Это поможет избежать очевидных ошибок.

### Составление сводных таблиц

Представленных данных хватит для того, чтобы ответить на вопросы заказчика. Однако, сперва необходимо визуализировать их подходящим для анализа образом.
#### Сводная таблица по наличию детей

In [46]:
# проверим качество данных ещё раз

df_clear['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

In [47]:
# построим сводную таблицу
# в качестве индекса укажем количество детей, 
# чтобы оценить совместное влияние этих факторов
# применим к значениям 'debt' функцию mean,
# чтобы посчитать среднее число возвратов
# сформировав процентную величину вероятности

df_children_pivot = df_clear.pivot_table(index=['children'], values='debt', aggfunc='mean')

# добавим столбец, содержащий разницу вероятности базового возрата 
# и рассчитанного, с учетом дополнительных критериев

df_children_pivot['total_impact'] = 0.088241 - df_children_pivot['debt']
df_children_pivot

debt  total_impact
children                        
0         0.075438      0.012803
1         0.092346     -0.004105
2         0.094542     -0.006301
3         0.081818      0.006423
4         0.097561     -0.009320
5         0.000000      0.088241

Цифры говорят сами за себя. Но к выводам мы перейдем в следующем разделе.

#### Сводная таблица по семейному положению

In [48]:
# проверим качество данных ещё раз

df_clear['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

In [49]:
# повторим процедуру построения сводной таблицы

df_family_pivot = df_clear.pivot_table(index=['family_status'], values='debt', aggfunc='mean')

# добавим столбец, содержащий разницу вероятности базового возрата 
# и рассчитанного, с учетом дополнительных критериев

df_family_pivot['total_impact'] = 0.088241 - df_family_pivot['debt']
df_family_pivot

debt  total_impact
family_status                                
в разводе              0.070648      0.017593
вдовец / вдова         0.066246      0.021995
гражданский брак       0.093130     -0.004889
женат / замужем        0.075606      0.012635
не женат / не замужем  0.097639     -0.009398

#### Сводная таблица по уровню дохода

In [50]:
# проверим качество данных ещё раз

df_clear['income_class'].unique()

array([4, 2, 3, 1, 5, 0], dtype=int64)

In [51]:
# повторим процедуру построения сводной таблицы

df_income_pivot = df_clear.pivot_table(index=['income_class'], values='debt', aggfunc='mean')

# добавим столбец, содержащий разницу вероятности базового возрата 
# и рассчитанного, с учетом дополнительных критериев

df_income_pivot['total_impact'] = 0.088241 - df_income_pivot['debt']
df_income_pivot 

debt  total_impact
income_class                        
0             0.069471      0.018770
1             0.088197      0.000044
2             0.086074      0.002167
3             0.083467      0.004774
4             0.070251      0.017990
5             0.071019      0.017222

#### Сводная таблица по целям кредитов

In [52]:
# проверим качество данных ещё раз

df_clear['purpose_class'].unique()

array(['недвижим', 'автомобил', 'образован', 'свадьб'], dtype=object)

In [53]:
# повторим процедуру построения сводной таблицы

df_purpose_pivot = df_clear.pivot_table(index=['purpose_class'], values='debt', aggfunc='mean')

# добавим столбец, содержащий разницу вероятности базового возрата 
# и рассчитанного, с учетом дополнительных критериев

df_purpose_pivot['total_impact'] = 0.088241 - df_purpose_pivot['debt']
df_purpose_pivot

debt  total_impact
purpose_class                        
автомобил      0.093480     -0.005239
недвижим       0.072551      0.015690
образован      0.092528     -0.004287
свадьб         0.079118      0.009123

Все расчеты и исследования произведены, время переходить к выводам.

## Выводы

Здесь и далее, мы будем использовать термин ВКП - Вероятность возврата Кредита без Просрочки. Базовый ВКП - ВКП, рассчитанный на основании общих данных, представленных заказчиком, без учета влияния каких-либо характеристик заемщика.

ВКП без просрочки равняется 8.8%. У нас отсутствует информация о порядке формирования данного значения. Вполне вероятно, что в эту цифру попадают даже те, у кого есть просрочка в один-два дня. В связи с чем   сложно сделать вывод о плачевности ситуации. Однако, именно этот факт необходимо держать в голове при проведении анализа и оценке его итогов. 


По результатам проведенного исследования, можно прийти к выводу, что ни одна из гипотез, заявленных к анализу заказчиком, не имеет хоть сколь сильного влияния на ВКП. Представленные категории могут повлиять на цель кредита, его размер, но не на итоговую просрочку. 


Более того, представленные данные больше подходят для создания модели, предсказывающей вероятность ВКП у каждого конкретного клиента. Особенно, в случае уточнения термина "просрочка" и разбитие его на временные периоды, исходя из специфики банковской деятельности. Кроме того, полученная в ходе исследования информация, помогла выявить несовершенство в формировании представленных данных, о чем, безусловно, стоит информировать заказчика.

Учитывая достаточно однородную структуру данных, мы можем посчитать усредненный процент влияния категорий, озвученных заказчиком, на возврат кредита в срок.

In [54]:
# напишем функцию, для подсчета значений, 
# заносимых в финальную таблицу

def totalizer(list_name, name, table):
    '''
    Функция totalizer служит для наполнения списков вычислениями.
    Она принимает в качестве параметра название списка (list_name),
    значение (name), которое добавляет в начало списка,
    таблицу (table) из которой берутся данные для вычисления.
    Функция возвращает значение (name), max() для базового и измененного, 
    min() для базового и измененного и mean() для измененного значений ВКП,
    создавая список из 6 элементов.
    '''
    list_name.append(name)
    list_name.append(table['debt'].max())
    list_name.append(table['debt'].min())
    list_name.append(table['total_impact'].max())
    list_name.append(table['total_impact'].min())
    list_name.append(table['total_impact'].mean())

# создадим списки, которые наполним функцией

children_total_impact_list = []
family_total_impact_list = []
income_total_impact_list = []
purpose_total_impact_list = []

# запустим функцию последовательно для каждого списка циклом

totalizer(children_total_impact_list, 'children', df_children_pivot)
totalizer(family_total_impact_list, 'family', df_family_pivot)
totalizer(income_total_impact_list, 'income', df_income_pivot)
totalizer(purpose_total_impact_list, 'purpose_class', df_purpose_pivot)

# создадим список списков, чтобы наполнить таблицу
# выведем его на экран, чтобы проверить работу функции

total_impact_list = [
    children_total_impact_list, family_total_impact_list,
    income_total_impact_list, purpose_total_impact_list
]

print(total_impact_list)

# соберем полученные данные в единую таблицу

impact_columns = [
    'impact_object', 'max_impact_value', 'min_impact_value', 
    'max_impact_delta', 'min_impact_delta', 'mean_impact_delta'
]

results = pd.DataFrame(data=total_impact_list, columns=impact_columns) 
results

[['children', 0.0975609756097561, 0.0, 0.088241, -0.009319975609756101, 0.014623436568536983], ['family', 0.09763948497854077, 0.06624605678233439, 0.02199494321766561, -0.009398484978540775, 0.007587227249472408], ['income', 0.08819714656290532, 0.06947100063734862, 0.018769999362651377, 4.385343709467637e-05, 0.01016102316288911], ['purpose_class', 0.09347978499649451, 0.07255139056831923, 0.015689609431680768, -0.005238784996494508, 0.0038218032881426438]]


impact_object  max_impact_value  min_impact_value  max_impact_delta  \
0       children          0.097561          0.000000          0.088241   
1         family          0.097639          0.066246          0.021995   
2         income          0.088197          0.069471          0.018770   
3  purpose_class          0.093480          0.072551          0.015690   

   min_impact_delta  mean_impact_delta  
0         -0.009320           0.014623  
1         -0.009398           0.007587  
2          0.000044           0.010161  
3         -0.005239           0.003822

**ИТОГИ**

Несмотря на предположение, что какие-либо факторы могут оказывать влияние на ВКП, данные не подтвердили эту гипотезу, продемонстрировав незначительные изменения, в зависимости от критериев заемщика. 


Наибольшее влияние на ВКП оказало **количество детей** - максимальное увеличение составляет почти до 9% (эта цифра, в целом, съедается на фоне не самого высокго базового ВКП группы). Примечательно, что наибольшее увеличение показано для тех, у кого имеется 5 детей. Это значение можно смело считать артефактом, не отражающим действительность, так как таковых клиентов в данных всего 9. Прямым подтверждением того, что дети являются обеспечением ВКП, является самый высокий из всех групп средний рост ВКП, который составил почти 2%. При этом заемщики без детей, в целом, демонстрируют лучшую динамику ВКП, чем с детьми. Количество детей не влияет на динамику ухудшения ВКП, сохраняя одинаковый прогресс для 2 и более детей (с учетом описанного нами артефакта для большого количества детей в семье).
**Рекомендации:** предпочтение составляют клиенты без детей (кто бы мог подумать). Влияние большого количества детей на ВКП изучить на представленных данных невозможно.


На втором месте по влиянию на ВКП оказывается **размер дохода** заемщика. Этот критерий демонстрирует увеличение максимального ВКП на 8.8%. Прежде всего, это связано с более высоким базовым ВКП у некоторых классов доходов. При этом, как ни парадоксально, стойкое увеличение ВКП показывают 4 и 5 группы, а также 0, с наименьшим доходом. Группы с 1 по 2 класс дохода демонстрирую ухудшение ВКП, с постепенным улучшением ситуации, начиная с 3 класса. Однако, не стоит присваивать этому выводу большой вес, так как среднее изменение ВКП 1%. Учитывая прямую корелляцию вероятности возврата кредита без просрочек, в зависимости от размера дохода - класс дохода обеспечивает наиболее стабильный прогноз ВКП.
**Рекомендации:** крайние значения классов дохода оказывают незначительное влияние на ВКП, предпочтение стоит отдавать людями с доходом 170 и более тысяч рублей (еще один "неожиданный" вывод).


На третьем месте по влиянию на ВКП находится **семейный статус** заемщика. Умеренная динамика роста крайних значений ВКП (2%), незначительное ухудшение базового ВКП (на тысячные процента). Интересно, что оценка ВКП с позиции семейного статуса приводит к снижению базового ВКП. При этом наивысшый положительный рост ВКП демонстрируют категории заемщиков состоящих в гражданском браке, а также одинокие. Важным нюансом является то, что разведенные демонстрируют гораздо более плохие параметры ВКП, чем не вступавшие в брак. Что, в принципе, обосновываются тяжестью процесса, а также алиментными платежами, уменьшающими реальный доход заемщика. 
**Рекомендации:** одинокий (но не разведенные) и состоящие в гражданском браке представляют приоритетную категорию заемщиков (при этом ВКП кредита на свадьбу для состоящих в гражданском браке демонстрирует самый худший ВКП 7,9% по сравнению с ближайшим 9,1% при сделках с недвижимостью). Вдоцвы и разведенные составляют худшую категорию заемщиков, имеющих стойкую тенденцию к снижению ВКП.


На последнем месте по изменению ВКП находится **цель кредита**. Кредиты на свадьбу и недвижимость имеют тенденцию к росту ВКП. Но при этом проигрывают ВКП при кредитовании на автомобили и образование. Эти цели обеспечивают большие крайние значения прироста ВКП, а медиана оставляют при этом почти без изменений, с ухудшением на десятые доли процента.
**Рекомендации:** кредиты на авто и образование тяготеют к более высокому ВКП. Кредиты на свадьбу и недвижимость, зачастую связаны с большими тратами и склонны к снижению ВКП.


Подводя итог, можно сказать, что характеристики заемщика могут оказывать влияние, но на фоне средних значений они практически не заметны. Возможно дополнительные критерии, представленные в дата-сете могут изменить картину ВКП, но они выходят за границы исследования.

## Обнаруженные проблемы

В ходе исследования были выявлены различные проблемы, устранение которых может привести к улучшению данных заказчика:
1. Отсутствует документация по числовым значениям в столбце 'family_status_id'. В ходе анализа нами был добавлен словарь.
2. Стаж, согласно документации, измеряется в днях, однако максимальное значение составляет 400 тысяч, а также имеет отрицательные значения. Это проблема может заключаться в системе формирования данных или человеческом факторе. В любом случае, стоит уточнить этот момент.
3. Длительность стажа и доход формируются в вещественных числах, что не имеет смысла.
4. Разброс дохода в текущих данных слишком большой, что не поддается анализу. Его необходимо классифицировать. Нами было предложено две модели классификации. Естественно, что зависить она должна от региона.
5. Аномальные значения в количестве детей также являются ошибкой.
6. Полное отсутствие единообразия классификации целей кредита может ухудшать работу внутренних служб заказчика, а также поиска необходимых данных. Нами была разработана базовая классификация, которая может быть взята за основу, но с уточнениями.